keps/sig-cluster-lifecycle/kubeadm/20190722-Advanced-configurations-with-kubeadm-(Kustomize).md

```diff
@@ -0,0 +1,365 @@
+“advanced configuration” already in place
+
+### Implementation Details
+
+This proposal explores as a first option for implementing Kubeadm 
+“advanced configurations“ the usage of Kustomize; please refer to
+[Declarative application management in Kubernetes](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/architecture/declarative-application-management.md)
+and [Kustomize KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cli/0008-kustomize.md)
+for background information about Kustomize.
+
+#### Kustomize integration with kubeadm
+
+By adopting Kustomize in the context of Kubeadm, this proposal assumes to:
+
+- Let kubeadm generate the static pod manifests as usual.
+- Use kubeadm generated artifacts as a starting point for applying patches
+  containing “advanced configurations”.
+
+This has some implications:
+
+1. The higher-level tools/users have to express “advanced configurations” using
+   one of the two alternative techniques supported by Kustomize - the [strategic
+   merge patch](https://github.com/kubernetes-sigs/kustomize/blob/master/docs/glossary.md#patchstrategicmerge)
+   and the [JSON patch](https://github.com/kubernetes-sigs/kustomize/blob/master/docs/glossary.md#patchjson6902) -.
+2. The higher-level tools/users have to provide patches before running kubeadm;
+   this point is going to be further discussed in the following paragraphs.
+3. Kubeadm is responsible for coordinating the execution of Kustomize within the
+   init/join/upgrade workflows
+4. as a consequence of the previous point, higher-level tools/users are not
+   requested to take care of defining `kustomization.yaml` files nor to define
+   a local folder structure.
+5. `patchesJson6902` can't be used without a `kustomization.yaml` file defining `target`
+   for a patch; this limitation is not considered blocking for starting implementation;
+   however a solution for this problem should be defined before graduating to beta
+   (e.g. require `kustomization.yaml` in case the users want to use `patchesJson6902`).
+
+Additionally, in order to simplify the first implementation of this KEP, this 
+proposal is going to assume that Kustomize patches for kubeadm are always defined
+specifically for the node where kubeadm is going to be executed.
+
+This point could be reconsidered in the future, by e.g. introducing cluster-wide
+patches and/or patches for a subset of nodes.
+
+The resulting workflow for using in kustomize will be the following
+
+1. Create a folder with some patches, e.g.
+
+```
+mkdir kubeadm-patches
+
+cat <<EOF >./kubeadm-patches/patch1.yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: kube-apiserver
+  namespace: kube-system
+{your kube-apiserver patch here}
+EOF
+
+cat <<EOF >./kubeadm-patches/patch2.yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: kube-controller-manager
+  namespace: kube-system
+{your kube-controller-manager patch here}
+EOF
```

Comment:
Should there also be an example here for a net new static pod?

Reply:
@detiber I'm not sure what do you mean with "net new static pod". 

This feature targets only the 4 static pods manifests (3 in case of external etcd) created by kubeadm in the `/etcd/kubernetes/manifests` folder